### PR TITLE
Pokemon list

### PR DESCRIPTION
**Pull request by jogracz**
I used FlatList with getMorePokemons function triggered when scrolled close to the bottom of content

### CODE REVIEW

File: api/pokemon.tsx [hunk 1]
```diff
@@ -11,10 +19,15 @@ export default async function getFavPokemon(favPokemonName) {
...
-  height: number;
-  base_experience: number;
-  sprites: [object];
+export async function getAllPokemons(offset, limit) {
+  try {
+    const res = await fetch(
+      `https://pokeapi.co/api/v2/pokemon?offset=${offset}&limit=${limit}`
+    );
+    const data = await res.json();
+    return data.results;
```
Comment (barsumek) [inline api/pokemon.tsx]: I think you could return it as `ReadonlyArray<Pokemon>`

File: screens/PokemonList.tsx [hunk 2]
```diff
@@ -1,29 +1,100 @@
...
-        onPress={() => navigation.navigate('NativeModule')}
-      /> */}
-    </View>
-  );
+import { StyleSheet, Text, View, FlatList, Dimensions } from 'react-native';
+import { getAllPokemons } from '../api/pokemon';
+
+const keyExtractor = ({ name }) => name;
+
+const PokemonList = () => {
```
Comment (barsumek) [inline screens/PokemonList.tsx:7]: Change to `function` after fixing the conflict with `master`

File: screens/PokemonList.tsx [hunk 3]
```diff
@@ -1,29 +1,100 @@
...
+  const [batchStart, setBatchStart] = useState(0);
+  const [loading, setLoading] = useState(true);
+  const [pokemons, setPokemons] = useState(null);
+
+  useEffect(() => {
+    const effect = async () => {
+      const allPokemons = await getAllPokemons(batchStart, step);
+      setPokemons(allPokemons);
+      setLoading(false);
+      setBatchStart(batchStart + step);
```
Comment (barsumek) [inline screens/PokemonList.tsx]: I think you shouldn't set new start if `getAllPokemons` returned `null`

File: screens/PokemonList.tsx [hunk 4]
```diff
@@ -1,29 +1,100 @@
...
+    );
+  };
+
+  const scrolledCloseToEnd = ({
+    layoutMeasurement,
+    contentOffset,
+    contentSize
+  }) => {
+    return (
+      layoutMeasurement.height + contentOffset.y >= contentSize.height - 200
```
Comment (barsumek) [inline screens/PokemonList.tsx]: why `200`? If it's a "magic number" then it's good to move it to `const` with appriopriate name :)

File: screens/PokemonList.tsx [hunk 5]
```diff
@@ -1,29 +1,100 @@
...
+      setLoading(false);
+      setBatchStart(batchStart + step);
+    };
+    effect();
+  }, []);
+
+  const renderItem = ({ item }) => {
+    //const matchRes = item.url.match('(pokemon\\/{1}(\\d+)\\/{1})', 'gi');
+
+    // this is more readable than regex
```
Comment (barsumek) [inline screens/PokemonList.tsx]: true 😄 but 👏 for the effort of using regex. I'd think about extracting it to a function `extractPokemonIdFromUrl`

File: screens/PokemonList.tsx [hunk 6]
```diff
@@ -1,29 +1,100 @@
...
+    contentOffset,
+    contentSize
+  }) => {
+    return (
+      layoutMeasurement.height + contentOffset.y >= contentSize.height - 200
+    );
+  };
+
+  const getMorePokemons = async () => {
+    const morePokemons = await getAllPokemons(batchStart, step);
```
Comment (barsumek) [inline screens/PokemonList.tsx:49]: what if it returns `null`?

File: screens/PokemonList.tsx [hunk 7]
```diff
@@ -1,29 +1,100 @@
...
+    return (
+      <View style={styles.container}>
+        <Text style={styles.header}>Pokemon List</Text>
+
+        <FlatList
+          style={styles.list}
+          data={pokemons}
+          renderItem={renderItem}
+          keyExtractor={keyExtractor}
+          onScroll={({ nativeEvent }) => {
```
Comment (barsumek) [inline screens/PokemonList.tsx]: I think there's `onEndReached` which you could use instead (but it's good that you played with `onScroll`! 💯)

File: package.json [hunk 8]
```diff
@@ -15,6 +15,7 @@
     "axios": "^0.19.2",
     "expo": "~36.0.2",
     "react": "~16.13.0",
+    "react-bottom-scroll-listener": "^3.0.0",
```
Comment (barsumek) [inline package.json]: I think you are not using it, are you? 😛

**Review by barsumek — COMMENTED** [review]
Some nits, but nice work 👏

File: screens/PokemonList.tsx (hunk 4 above)
Comment (jogracz) [reply screens/PokemonList.tsx]: yeah, it was 1 at first but I wanted the new content to load before we reach the bottom

File: package.json (hunk 8 above)
Comment (jogracz) [reply package.json]: I found it and tried to use it here but didn't work. Is it just for plain react web apps?

File: screens/PokemonList.tsx (hunk 7 above)
Comment (jogracz) [reply screens/PokemonList.tsx]: Apparently there's a problem with onEndReached. It gets called on render :/

File: package.json (hunk 8 above)
Comment (barsumek) [reply package.json]: Yep, just React from what I see. I added this comment as if you're not using it then there's no point in keeping it in `package.json` -> please remove it and run `npm install` again :)

File: screens/PokemonList.tsx (hunk 7 above)
Comment (barsumek) [reply screens/PokemonList.tsx]: Could you elaborate what the problem is (here or on Slack)? It's quite "normal" to use it for refetching :) there's also `onEndReachedThreshold` for better control when exactly it's called

File: api/pokemon.tsx [hunk 9]
```diff
@@ -4,9 +4,12 @@ interface Pokemon {
   height: number;
   base_experience: number;
   sprites: [object];
+  url: string;
 }
 
-export default async function getFavPokemon(favPokemonName) {
+export default async function getFavPokemon(
+  favPokemonName: string
+): Promise<Pokemon> {
```
Comment (barsumek) [inline api/pokemon.tsx]: I think it should be `Promise<Pokemon | null>` -> is TypeScript happy with the current approach?

File: api/pokemon.tsx [hunk 10]
```diff
@@ -19,7 +22,10 @@ export default async function getFavPokemon(favPokemonName) {
   }
 }
 
-export async function getAllPokemons(offset, limit) {
+export async function getAllPokemons(
+  offset: number,
+  limit: number
+): Promise<ReadonlyArray<Pokemon>> {
```
Comment (barsumek) [inline api/pokemon.tsx]: same here

File: tsconfig.json [hunk 11]
```diff
@@ -6,6 +6,7 @@
     "moduleResolution": "node",
     "noEmit": true,
     "skipLibCheck": true,
-    "resolveJsonModule": true
+    "resolveJsonModule": true,
+    "strict": true
```
Comment (barsumek) [inline tsconfig.json:10]: you will probably have to fix other files too, but let's do it one by one 😄 nice work!

**Review by barsumek — APPROVED** [review]
🥇

**Pull request merged**